### PR TITLE
Implement long to int casts with overflow for x86

### PR DESCRIPTION
**Pull request by mikedn**
Contributes to #4169
Replaces #6526

### CODE REVIEW

**Comment by mikedn** [comment]
The generated overflow checks are practically identical to the ones generated by JIT32. There's an alternative implementation that trades a temp register for less branches:

long -> int cast:
```
mov tmp, lo
sar tmp, 31
cmp tmp, hi ; all upper 33 bits must be all 0 or all 1
jne overflow
```
ulong -> int cast:
```
mov tmp, lo
sar tmp, 31
or tmp, hi  ; all upper 33 bits must be all 0
jne overflow
```

It's tempting to get rid of the branches but I'm not sure if the increased use of registers is a good thing so I used the JIT32 approach.

File: src/jit/codegenxarch.cpp [hunk 1]
```diff
@@ -7580,6 +7573,93 @@ void CodeGen::genSetRegToCond(regNumber dstReg, GenTreePtr tree)
     }
 }
 
+#if !defined(_TARGET_64BIT_)
+//------------------------------------------------------------------------
+// genIntToIntCast: Generate code for long to int casts on x86.
```
Comment (pgavlin) [inline src/jit/codegenxarch.cpp]: Should be `genLongToIntCast`

File: src/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -7580,6 +7573,93 @@ void CodeGen::genSetRegToCond(regNumber dstReg, GenTreePtr tree)
...
+// Assumptions:
+//    The cast node and its sources (via GT_LONG) must have been assigned registers.
+//    The destination cannot be a floating point type or a small integer type.
+//
+void CodeGen::genLongToIntCast(GenTree* cast)
+{
+    assert(cast->OperGet() == GT_CAST);
+
+    GenTree* src = cast->gtGetOp1();
+    noway_assert(src->OperGet() == GT_LONG);
```
Comment (pgavlin) [inline src/jit/codegenxarch.cpp]: IMO this can simply be an `assert`.

**Comment by pgavlin** [comment]
LGTM

**Comment by pgavlin** [comment]
@dotnet/jit-contrib PTAL

**Comment by adiaaida** [comment]
LGTM too.

File: src/jit/codegenxarch.cpp (hunk 1 above)
Comment (mikedn) [inline src/jit/codegenxarch.cpp]: Oops. I guess I'll send another PR to fix this.

File: src/jit/codegenxarch.cpp (hunk 2 above)
Comment (mikedn) [inline src/jit/codegenxarch.cpp]: It certainly can, it just seems to me that we use `noway_assert` more often than `assert` for this particular check.

**Pull request merged**